### PR TITLE
Improve operator lock handling by releasing locks once cluster is reconciled

### PR DESCRIPTION
**Pull request by johscheuer**
# Description

*Please include a summary of the change and which issue is addressed. If this change resolves an issue, please include the issue number in the description.*

## Type of change

*Please select one of the options below.*

- New feature (non-breaking change which adds functionality)

## Discussion

Otherwise we have wait until the current lock expires. The default lock duration is 10 minutes, so waiting until the lock expires will slow down operations for HA clusters.

## Testing

Ran the unit tests (but they are not yet traversing the code path) and ran some e2e tests.

I will try to see if we can add some additional unit tests for this.

## Documentation

Will be updated once the changes look good (to make sure we document the right approach).

## Follow-up

-

### CODE REVIEW

File: controllers/bounce_processes.go [hunk 1]
```diff
@@ -251,27 +253,29 @@ func getProcessesReadyForRestart(logger logr.Logger, cluster *fdbv1beta2.Foundat
...
-	}
+	if cluster.IsBeingUpgradedWithVersionIncompatibleVersion() {
+		// If we upgrade the cluster wait until all processes are ready for the restart. We don't want to block the restart
+		// if some processes are already upgraded e.g. in the case of version compatible upgrades and we also don't want to
+		// block the restart command if a process is missing longer than the specified GetIgnoreMissingProcessesSeconds.
+		// Those checks should ensure we only run the restart command if all processes that have to be restarted and are connected
+		// to cluster are ready to be restarted.
+		expectedProcesses := counts.Total() - missingProcesses + markedForRemoval
+		// If more than one storage server per Pod is running we have to account for this. In this case we have to add the
+		// additional storage processes.
```
Comment (sbodagala) [inline controllers/bounce_processes.go:264]: It would be nice if we could push this logic into "GetProcessCountsWithDefaults()".

**Review by sbodagala — APPROVED** [review]

**Comment by johscheuer** [comment]
> ### Result of fdb-kubernetes-operator-pr on Linux CentOS 7
> * Commit ID: [5d2a656](https://github.com/FoundationDB/fdb-kubernetes-operator/commit/5d2a65639e93249a438168feab3df8c618e4eb1b)
> * Duration 2:44:18
> * Result: ❌ **FAILED**
> * Error: `Error while executing command: if $fail_test; then exit 1; fi. Reason: exit status 1`
> * [Build Log](https://d1e0l78xbkh3xa.cloudfront.net/fdb-kubernetes-operator/fdb-kubernetes-operator-pr/9ebf93f1-15aa-45ac-84e0-c8bd879aaa38.gz) terminal output (available for 30 days)
> * [Build Workspace](https://d1e0l78xbkh3xa.cloudfront.net/fdb-kubernetes-operator-pr/9ebf93f1-15aa-45ac-84e0-c8bd879aaa38/workspace.zip) zip file of the working directory (available for 30 days)

The failures are unrelated to this PR. Those are tests that are known to be flaky and must be fixed.

File: controllers/bounce_processes.go (hunk 1 above)
Comment (johscheuer) [reply controllers/bounce_processes.go:264]: I create an issue to look into this. We cannot move the code as it is directly into the `GetProcessCountsWithDefaults` method as this method is used in different places to get the number of process groups, so probably we need to implement a new method.

Comment (johscheuer) [reply controllers/bounce_processes.go:264]: https://github.com/FoundationDB/fdb-kubernetes-operator/issues/1804

**Pull request merged**